python/datasources/bjs.py

```diff
@@ -0,0 +1,393 @@
+    df = keep_only_states(df)
+
+    df = df.append(source_df_territories)
+    df[std_col.ALL_VALUE] = df[std_col.ALL_VALUE].combine_first(
+        df[Race.ALL.value])
+    df = df.drop(columns=[Race.ALL.value])
+
+    df = cols_to_rows(
+        df, BJS_SEX_GROUPS, std_col.SEX_COL, RAW_COL)
+
+    df = dataset_utils.merge_fips_codes(df)
+
+    df = dataset_utils.merge_pop_numbers(
+        df, std_col.SEX_COL, "state")
+
+    df[std_col.POPULATION_PCT_COL] = df[std_col.POPULATION_PCT_COL].astype(
+        float)
+
+    df[PER_100K_COL] = df.apply(calc_per_100k, axis="columns")
+
+    return df
+
+
+def make_prison_state_age_df(source_df_juveniles, source_df_totals, source_df_territories):
+
+    source_df_juveniles = source_df_juveniles[source_df_juveniles[std_col.STATE_NAME_COL]
+                                              != constants.US_NAME]
+
+    source_df_juveniles = source_df_juveniles.rename(columns={RAW_COL: '0-17'})
+
+    source_df_juveniles = source_df_juveniles.drop(columns=[std_col.AGE_COL])
+
+    source_df_totals = keep_only_states(source_df_totals)
+    source_df_totals = source_df_totals[[
+        std_col.STATE_NAME_COL, std_col.ALL_VALUE]]
+
+    df = pd.merge(source_df_juveniles, source_df_totals,
+                  on=std_col.STATE_NAME_COL)
+
+    df = df.append(source_df_territories)
+
+    df[std_col.ALL_VALUE] = df[std_col.ALL_VALUE].combine_first(
+        df[Race.ALL.value])
+    df = df.drop(columns=[Race.ALL.value])
+
+    df["18+"] = df[std_col.ALL_VALUE] - df['0-17']
+
+    df = cols_to_rows(df, BJS_AGE_GROUPS_JUV_ADULT, std_col.AGE_COL, RAW_COL)
+
+    df = dataset_utils.merge_fips_codes(df)
+    df = dataset_utils.merge_pop_numbers(
+        df, std_col.AGE_COL, "state")
+    df[std_col.POPULATION_PCT_COL] = df[std_col.POPULATION_PCT_COL].astype(
+        float)
+
+    df[PER_100K_COL] = df.apply(calc_per_100k, axis="columns")
+
+    return df
+
+
+def post_process(df, breakdown, geo):
+    """Merge the population data and then do all needed calculations with it.
+       Returns a dataframe ready for the frontend.
+
+       df: Dataframe with all the raw data.
+       breakdown: demographic breakdown (race, sex, age)
+       geo: geographic level (national, state)
+    """
+
+    df = dataset_utils.generate_pct_share_col(
+        df, RAW_COL, PCT_SHARE_COL, breakdown, std_col.ALL_VALUE)
+
+    df = df.drop(columns=[std_col.POPULATION_COL, RAW_COL])
+
+    return df
+
+
+def estimate_total(row, condition_name_per_100k):
+    """Returns an estimate of the total number of people with a given condition.
+
+       condition_name_per_100k: string column name of the condition per_100k to estimate the total of"""
+
+    if (pd.isna(row[condition_name_per_100k]) or
+        pd.isna(row[std_col.POPULATION_COL]) or
+            int(row[std_col.POPULATION_COL]) == 0):
+        return None
+
+    return round((float(row[condition_name_per_100k]) / 100_000) * float(row[std_col.POPULATION_COL]))
```

Comment:
I had temporarily just copied this from your code, but since this is being used in several data sources can I move this into the shared ingestion dataset_utils ?

Reply:
@joshzarrabi